lib/membrane/core/element/demand_handler.ex

```diff
@@ -173,6 +116,120 @@ defmodule Membrane.Core.Element.DemandHandler do
     end
   end
 
+  @doc """
+  Decreases the demand on the output by the size of outgoing buffers. Checks for the toilet
+  overflow if the toilet is enabled.
+  """
+  @spec handle_outgoing_buffers(
+          Pad.ref_t(),
+          PadModel.pad_data_t(),
+          [Buffer.t()],
+          State.t()
+        ) :: State.t()
+  def handle_outgoing_buffers(pad_ref, %{mode: :pull} = data, buffers, state) do
+    %{other_demand_unit: other_demand_unit, demand: demand} = data
+    buf_size = Buffer.Metric.from_unit(other_demand_unit).buffers_size(buffers)
+    PadModel.set_data!(state, pad_ref, :demand, demand - buf_size)
+  end
+
+  def handle_outgoing_buffers(_pad_ref, %{mode: :push, toilet: toilet} = data, buffers, state)
+      when toilet != nil do
+    %{other_demand_unit: other_demand_unit, pid: pid} = data
+    buf_size = Buffer.Metric.from_unit(other_demand_unit).buffers_size(buffers)
+    toilet_size = :atomics.add_get(toilet, 1, buf_size)
+
+    if toilet_size > 200 do
+      Membrane.Logger.debug_verbose(~S"""
+      Toilet overflow
+
+                   ` ' `
+               .'''. ' .'''.
+                 .. ' ' ..
+                '  '.'.'  '
+                .'''.'.'''.
+               ' .''.'.''. '
+             ;------ ' ------;
+             | ~~ .--'--//   |
+             |   /   '   \   |
+             |  /    '    \  |
+             |  |    '    |  |  ,----.
+             |   \ , ' , /   | =|____|=
+             '---,###'###,---'  (---(
+                /##  '  ##\      )---)
+                |##, ' ,##|     (---(
+                 \'#####'/       `---`
+                  \`"#"`/
+                   |`"`|
+                 .-|   |-.
+            jgs /  '   '  \
+                '---------'
+      """)
+
+      Membrane.Logger.error("""
+      Toilet overflow.
+
+      Reached the size of #{inspect(toilet_size)},
+      which is above fail level when storing data from output working in push mode.
+      To have control over amount of buffers being produced, consider using pull mode.
+      """)
+
+      Process.exit(pid, :kill)
+    end
+
+    state
+  end
+
+  def handle_outgoing_buffers(_pad_ref, _pad_data, _buffers, state) do
+    state
+  end
+
+  defp update_demand(pad_ref, size, state) when is_integer(size) do
+    PadModel.set_data!(state, pad_ref, :demand, size)
+  end
+
+  defp update_demand(pad_ref, size_fun, state) when is_function(size_fun) do
+    PadModel.update_data!(
+      state,
+      pad_ref,
+      :demand,
+      fn demand ->
+        new_demand = size_fun.(demand)
+
+        if new_demand < 0 do
+          raise Membrane.ElementError,
+                "Demand altering function requested negative demand on pad #{inspect(pad_ref)} in #{state.module}"
+        end
+
+        new_demand
+      end
+    )
```

Comment:
Maybe get&set? this seems like something that might have impact on performance but most importantly it's shorter and less indented :wink: 
```suggestion
    demand = PadModel.get_data!(state, pad_ref, :demand)
    new_demand = size_fun.(demand)

    if new_demand < 0 do
      raise Membrane.ElementError,
               "Demand altering function requested negative demand on pad #{inspect(pad_ref)} in #{state.module}"
    end

    PadModel.set_data!(state, pad_ref, :demand, new_demand)
```